Implement some more checks in `ptr_guaranteed_cmp`.

zachs18 opened the pull request:
* Pointers with different residues modulo their allocations' least common alignment are never equal.
* Pointers to the same static allocation are equal if and only if they have the same offset.
* Pointers to different non-zero-sized static allocations are unequal if both point within their allocation, and not on opposite ends.

Tracking issue for `const_raw_ptr_comparison`: <https://github.com/rust-lang/rust/issues/53020>

This should not affect `is_null`, the only usage of this intrinsic on stable.

Closes https://github.com/rust-lang/rust/issues/144584

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

zachs18 committed:
Implement some more checks for `ptr_guaranteed_cmp` in consteval:

Pointers with different residues modulo their least common allocation alignment are never equal.
Pointers to the same static allocation are equal if and only if they have the same offset.
Strictly in-bounds (in-bounds and not one-past-the-end) pointers to different static allocations are always unequal.
A pointer cannot be equal to an integer if `ptr-int` cannot be null.

Also adds more tests for `ptr_guaranteed_cmp`.

Co-authored-by: Ralf Jung <post@ralfj.de>

RalfJung commented:
If you just want to fix https://github.com/rust-lang/rust/issues/144584, please reduce the PR to only affect the case where both pointers are based on the same static.

davidtwco commented:
r? @RalfJung

RalfJung reviewed (commented):
That's an impressive test, thanks a lot for writing it!

zachs18 commented on tests/ui/const-ptr/guaranteed_cmp.rs:
Ah, nice! There's an existing `tests/ui/consts/ptr_comparison.rs`, so I'll combine this with that.

RalfJung commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -289,13 +289,91 @@ impl<'tcx> CompileTimeInterpCx<'tcx> {
+                            "same static item DefId had two different AllocIds? {a_allocid:?} != {b_allocid:?}, {a_did:?} == {b_did:?}"
+                        );
+
+                        if a_info.size == Size::ZERO || b_info.size == Size::ZERO {
+                            // One or both allocations is zero-sized, so we can't know if the
+                            // pointers are (in)equal.
+                            // FIXME: Can zero-sized static be "within" non-zero-sized statics?
+                            // Conservatively we say yes, since that doesn't cause them to
+                            // "overlap" any bytes, but if not, then we could delete this branch
+                            // and have the other branches handle ZST allocations.
+                            2
```
Ah, right... we had that discussion when adding the reference text you mentioned, and IMO we shouldn't allow it, but I don't think we reached consensus on that.

zachs18 commented:
> > and adds alignment checks when comparing pointers with integer-valued pointers.
> 
> I would have preferred if you didn't extend the scope of the PR after I already did most of my review. Now I'll have to re-review, so back on the queue it goes.

This is a separate commit. If you prefer I can split this into a separate PR.

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17108579709/job/48523903194) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/48523903194) [(plain)](https://github.com/rust-lang/rust/commit/580074a4961272f36b2425e1793689b93a751a1f/checks/48523903194/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |
402 |                         // immutable statics can overlap with other kinds of allocations somtimes.
    |                                                                                          ^^^^^^^^
    |
some tidy checks failed
tidy error: checks with external tool 'typos' failed
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1583:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1225:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:26
  local time: Wed Aug 20 19:47:44 UTC 2025
  network time: Wed, 20 Aug 2025 19:47:44 GMT
##[error]Process completed with exit code 1.
```

</details>

zachs18 commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -287,15 +287,130 @@ impl<'tcx> CompileTimeInterpCx<'tcx> {
+                            // the other branches would already handle ZST allocations correctly.
+                            2
+                        } else if a_offset > a_info.size || b_offset > b_info.size {
+                            // One or both pointers are out of bounds of their allocation,
+                            // so conservatively say we don't know.
+                            // FIXME: we could reason about how far out of bounds the pointers are,
+                            // e.g. two pointers cannot be equal if them being equal would require
+                            // their statics to overlap.
+                            2
+                        } else if (a_offset == Size::ZERO && b_offset == b_info.size)
+                            || (a_offset == a_info.size && b_offset == Size::ZERO)
+                        {
+                            // The pointers are on opposite ends of different allocations, we
+                            // cannot know if they are equal, since the allocations may end up
+                            // adjacent at runtime.
+                            2
+                        } else {
+                            // The pointers are within (or one past the end of) different
+                            // non-zero-sized static allocations, and they are not at oppotiste
+                            // ends, so we know they are not equal because non-zero-sized statics
+                            // cannot overlap or be deduplicated, as per
+                            // https://doc.rust-lang.org/nightly/reference/items/static-items.html#r-items.static.intro
+                            // (non-deduplication), and
+                            // https://doc.rust-lang.org/nightly/reference/items/static-items.html#r-items.static.storage-disjointness
+                            // (non-overlapping)
+                            0
+                        }
+                    } else {
+                        // Even if one of them is a static, as per https://doc.rust-lang.org/nightly/reference/items/static-items.html#r-items.static.storage-disjointness
+                        // immutable statics can overlap with other kinds of allocations sometimes.
+                        // FIXME: We could be more decisive for (non-zero-sized) mutable statics,
+                        // which cannot overlap with other kinds of allocations.
+                        // `GlobalAlloc::{Memory, Function, Vtable}` can at least be deduplicated with
+                        // the same kind, so comparing two of the same kind of those should return 2.
+                        // `GlobalAlloc::TypeId` exists mostly to prevent consteval from comparing
+                        // `TypeId`s, so comparing two of those should always return 2.
+                        // FIXME: Can we determine any other cases?
+                        2
+                    }
+                }
```
Yes, that would work. I think I did it this way to have the comments be more precise to which cases they cover, but the comments could definitely be combined, especially for the cases that can both be duplicated and de-duplicated.

zachs18 replied:
Latest push does this.

cjgillot commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -287,15 +287,130 @@ impl<'tcx> CompileTimeInterpCx<'tcx> {
+                            "same static item DefId had two different AllocIds? {a_allocid:?} != {b_allocid:?}, {a_did:?} == {b_did:?}"
+                        );
+
+                        if a_info.size == Size::ZERO || b_info.size == Size::ZERO {
+                            // One or both allocations is zero-sized, so we can't know if the
+                            // pointers are (in)equal.
+                            // FIXME: Can zero-sized static be "within" non-zero-sized statics?
+                            // Conservatively we say yes, since that doesn't cause them to
+                            // "overlap" any bytes, but if not, then we could delete this branch;
+                            // the other branches would already handle ZST allocations correctly.
+                            2
+                        } else if a_offset > a_info.size || b_offset > b_info.size {
+                            // One or both pointers are out of bounds of their allocation,
+                            // so conservatively say we don't know.
+                            // FIXME: we could reason about how far out of bounds the pointers are,
+                            // e.g. two pointers cannot be equal if them being equal would require
+                            // their statics to overlap.
+                            2
+                        } else if (a_offset == Size::ZERO && b_offset == b_info.size)
+                            || (a_offset == a_info.size && b_offset == Size::ZERO)
+                        {
+                            // The pointers are on opposite ends of different allocations, we
+                            // cannot know if they are equal, since the allocations may end up
+                            // adjacent at runtime.
+                            2
+                        } else {
+                            // The pointers are within (or one past the end of) different
+                            // non-zero-sized static allocations, and they are not at oppotiste
+                            // ends, so we know they are not equal because non-zero-sized statics
+                            // cannot overlap or be deduplicated, as per
+                            // https://doc.rust-lang.org/nightly/reference/items/static-items.html#r-items.static.intro
+                            // (non-deduplication), and
+                            // https://doc.rust-lang.org/nightly/reference/items/static-items.html#r-items.static.storage-disjointness
+                            // (non-overlapping)
+                            0
+                        }
```
Please try to do as simple as possible. As a rule of thumb, it's better to list cases where we ensure stuff and return 0 or 1, and have the fallback to "I don't know". A "both inbound then 1", else "some special case then 1" else 2 is the best way.

zachs18 replied:
Latest push only returns unequal if both pointers are strictly in-bounds of their own allocations for simplicity.

RalfJung commented on tests/ui/consts/ptr_comparisons.rs:
```diff
@@ -1,43 +1,184 @@
 
 // aside from 0, these pointers might end up pretty much anywhere.
-check!(!, FOO as *const _, 1); // this one could be `ne` by taking into account alignment
-check!(!, FOO as *const _, 1024);
+do_test!(&A, align_of::<T>() as *const u8, None);
+do_test!(&A, 1 as *const u8, Some(false)); // this one takes into account alignment, so we know that
```
This grouping no longer makes sense... you first say "can end up anywhere" but then still have a `Some` test.

Please also ensure that `(&raw const A).wrapping_byte_add(1)` might be equal to 1. (I guess technically we could exclude that since statics cannot live at address 0....)

zachs18 replied:
> Please also ensure that `(&raw const A).wrapping_byte_add(1)` might be equal to 1. (I guess technically we could exclude that since statics cannot live at address 0.... but at some point we have to stop adding more cases to this PR.^^)

(The change added in 2a24220ef5d90e90fa216a60d0613a4500f9a169 makes this known to never occur)

zachs18 commented on compiler/rustc_const_eval/src/const_eval/machine.rs:
```diff
@@ -287,15 +287,130 @@ impl<'tcx> CompileTimeInterpCx<'tcx> {
+            // Other ways of comparing integers and pointers can never be known for sure,
+            // except for alignment, e.g. `1 as *const _` can never be equal to an even offset
+            // in an `align(2)` allocation.
+            (Scalar::Int(int), Scalar::Ptr(ptr, _)) | (Scalar::Ptr(ptr, _), Scalar::Int(int)) => {
```
Added in 2a24220ef5d90e90fa216a60d0613a4500f9a169

RalfJung reviewed (commented):
Okay, let's go with this logic for now. :)
Just some comment nits.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/144885/07b7dc90ee4df5815dbb91ef8e98cb93571230f5..2a24220ef5d90e90fa216a60d0613a4500f9a169)*

jhpratt commented:
@bors r-

https://github.com/rust-lang/rust/pull/145804#issuecomment-3217734839

zachs18 committed:
Remove arm::t32-specific function pointer test.

All function pointers are currently treated as unaligned anyway;
any change implementing function pointer alignment during consteval should add
tests that it works properly on arm::t32 functions.

zachs18 commented:
Latest commit removes the failing `#[instruction_set(arm::t32)]` aligned fn pointer test (and updates the comment in the test). Since https://github.com/rust-lang/rust/pull/144706 all fn pointers have no alignment in consteval anyway (regardless of instruction set or compilation target), so I didn't see much point trying to fix the test to accurately detect when `#[instruction_set(arm::t32)]` is supported when it isn't even testing anything different than the normal aligned `fn` pointer test at the moment.

@rustbot ready

RalfJung commented:
@bors r+

bors commented:
:pushpin: Commit 25afbbc9814dba4ad6f06b6d1bf14e92851ccd32 has been approved by `RalfJung`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=RalfJung 25afbbc9814dba4ad6f06b6d1bf14e92851ccd32 -->
<!-- homu: {"type":"Approved","sha":"25afbbc9814dba4ad6f06b6d1bf14e92851ccd32","approver":"RalfJung","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
